src/Enforcers/ValidationEnforcer.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+namespace PDGA\DataObjects\Enforcers;
+
+use \ReflectionClass;
+
+use PDGA\DataObjects\Validators\BoolValidator;
+use PDGA\DataObjects\Validators\DateValidator;
+use PDGA\DataObjects\Validators\IntValidator;
+use PDGA\DataObjects\Validators\NotNullValidator;
+use PDGA\DataObjects\Validators\StringValidator;
+use PDGA\Exception\ValidationListException;
+
+class ValidationEnforcer
+{
+    /**
+     * Validates the properties of an object against a class definition.
+     *
+     * @param array|object $object - The values to validate.
+     * @param string $className - The name of the class to validate against.
+     *
+     * @throws ValidationListException - Throws a ValidationListException which
+     * includes a cumulative list of errors encountered during validation for each
+     * property on the object.
+     */
+    public function validate(mixed $object, string $className): void
+    {
+
+        $validators = [
+            "int"      => new IntValidator(),
+            "string"   => new StringValidator(),
+            "bool"     => new BoolValidator(),
+            "DateTime" => new DateValidator(),
+            "null"     => new NotNullValidator(),
+        ];
+
+        $arr = is_array($object) ? $object : (array) $object;
+        $validationErrors = new ValidationListException();
+
+        $ref   = new ReflectionClass($className);
+        $props = $ref->getProperties();
+
+        //For each property defined on the class.
+        foreach($props as $prop)
+        {
+            $propName    = $prop->getName();
+            $propType    = $prop->getType()->getName();
+            $propAttrs   = $prop->getAttributes();
+            $propCanNull = $prop->getType()->allowsNull();
+
+            //If the class property is passed in with the object.
+            if ($this->propIsDefined($arr, $propName))
+            {
+                //If the property is not null and a validator exists for that type make sure the type is correct.
+                $validator = array_key_exists($propType, $validators) ? $validators[$propType] : null;
+                if ($this->propIsNotNull($arr, $propName) && !is_null($validator))
+                {
+                    if (!$validator->validate($arr[$propName]))
+                    {
+                        $validationErrors->addError($validator->getErrorMessage($propName), $propName);
+                    }
+                }
+                //If the property is null make sure it is allowed to be null.
+                else if (!$propCanNull)
```

Comment:
This logic looks off.  If the property has a value, there is no type validator, and the property is not nullable, this branch will be true, right?  Consider a float, for example, since we don't have a FloatValidator.

```
class FooDataObject
{
    public float someNumber; // Float that can't be null.
}

$v = new ValidationEnforcer();
$v->validate(['someNumber' => .42'], FooDataObject::class);
```

Wouldn't that give an error about `someNumber` being null, even though it has a value?

Reply:
Right, the second condition in the if statement makes this a little more complicated.  Will update the else if to include a check for null value.